### PR TITLE
feat: added support fuzzy word searching for X party modules

### PR DESCRIPTION
**Pull request by sandeshdamkondwar**
Added support for fuzzy word searching in X party modules list. 

#1170
Screenshot 2020-06-19 at 2 52 26 PM
Screenshot 2020-06-19 at 2 52 12 PM
Screenshot 2020-06-19 at 2 46 26 PM
Screenshot 2020-06-19 at 2 53 07 PM

### CODE REVIEW

File: pages/x/index.tsx [hunk 1]
```diff
@@ -17,18 +17,22 @@ const ThirdPartyRegistryList = () => {
     setQuery(event.target.value);
   }
 
-  const list = useMemo(
+  // Sorting the list only if entries are changed
+  const list = useMemo(() => {
```
Comment (rjoydip) [inline pages/x/index.tsx:21]: You don't need to use `filter` separately. You could do something like below. 

```ts
const list = useMemo(
    () =>
      Object.keys(entries)
        .filter((name) => {
          const queryReg = new RegExp(
            query.split(" ").reduce((a, b) => a + ".*" + b)
          );
          return (
            name.match(queryReg) || (entries[name].desc ?? "").match(queryReg)
          );
        })
        .sort((nameA, nameB) => nameA.localeCompare(nameB)),
    [entries, query]
  )
```

File: pages/x/index.tsx [hunk 2]
```diff
@@ -17,18 +17,22 @@ const ThirdPartyRegistryList = () => {
...
-          (name) =>
-            name.toLowerCase().includes(query.toLowerCase()) ||
-            (entries[name].desc ?? "")
-              .toLowerCase()
-              .includes(query.toLowerCase())
-        )
-        .sort((nameA, nameB) => nameA.localeCompare(nameB)),
-    [entries, query]
+      list.filter((name) => {
+        const queryReg = new RegExp(query.split(" ").join(".*?"), "i");
```
Comment (rjoydip) [inline pages/x/index.tsx:31]: > The regex itself can be optimised: .* initially grabs as much as possible, then retracts only if the rest doesn't match. This takes a larger than neccessary amount of backtracking to find out. Instead, you could use a reluctant quantifier: .*?. This attempts to match as little as possible. If a match is found, it is found without backtracking. However, it still does a lot of backtracking if there is no match. We can do better: a[^b]*b (at this point, it doesn't really matter if the quantifier is greedy). A possessive quantifier (a[^b]*+b) would be even better, but javascript doesn't support these. Using a character class deprives us of a join, but see the next point.

For more details [stackexchange](https://codereview.stackexchange.com/a/23905)
For benchmark [fuzzy-match](https://jsperf.com/fuzzy-match)

> Note: Search will be faster

**Comment by lucacasonato** [comment]
I think we should disable search for queries < 3 chars. On slow Android devices searching is pretty painful otherwise.

**Comment by rjoydip** [comment]
> I think we should disable search for queries < 3 chars. On slow Android devices searching is pretty painful otherwise.

Correct this will be good for mobile devices.

File: pages/x/index.tsx (hunk 1 above)
Comment (sandeshdamkondwar) [reply pages/x/index.tsx:21]: Does it make sense to sort every time query changes? Sorting algorithms are expensive compared to filter. And sorting is one time operation when entries are changed. That's the idea behind separating.

**Comment by sandeshdamkondwar** [comment]
> I think we should disable search for queries < 3 chars. On slow Android devices searching is pretty painful otherwise.

Agree, will update the PR.

**Comment by sandeshdamkondwar** [comment]
> > I think we should disable search for queries < 3 chars. On slow Android devices searching is pretty painful otherwise.
> 
> Agree, will update the PR.



> I think we should disable search for queries < 3 chars. On slow Android devices searching is pretty painful otherwise.

Gave a thought about this, not completely agree with this. People will search for terms like AB, API, URL, URI will end up in not finding what they want. Those terms generally goes inside description. Let's keep this way for now.

File: pages/x/index.tsx (hunk 2 above)
Comment (sandeshdamkondwar) [reply pages/x/index.tsx:31]: Both greedy a[^b]*b and possessive a[^b]*+b regexes are not working mentioned in [stackexchange link](https://codereview.stackexchange.com/a/23905).

`
const queryWords = query.split(" ").filter((word) => word.length);
queryWords.reduce((a, b) => {
        if (a === "") return b;
        return a + "[^" + b + "]*" + b;
}, "")
`

Tried here:
https://regex101.com/r/spw1H6/1
https://regex101.com/r/ecgzzN/1

@rjoydip  Do you have any solution for supporting backtracking?

**Comment by lucacasonato** [comment]
> Gave a thought about this, not completely agree with this point. People will search for terms like AB, API, URL, URI will end up in not finding what they want. Those terms generally goes inside description. Let's keep this way for now.

Terms like `API` are 3 characters. I was talking about queries < 3 not <=3: `"API".length < 3 == false`.

About things like `AB` => these will match so many modules anyway because it is just part of a word, that it doesn't make any sense to search for this anyway.

File: pages/x/index.tsx [hunk 3]
```diff
@@ -17,18 +17,22 @@ const ThirdPartyRegistryList = () => {
     setQuery(event.target.value);
   }
 
-  const list = useMemo(
+  // Sorting the list only if entries are changed
```
Comment (lucacasonato) [inline pages/x/index.tsx:20]: Sorting shouldn't actually be needed because `database.json` is already in alphabetical order.

Comment (sandeshdamkondwar) [reply pages/x/index.tsx:20]: @lucacasonato  Is there any automated job to update the list in database.json?

I don't see any. Devs might end up in writing anywhere in the list. 

Screenshot 2020-06-20 at 6 10 38 AM

Comment (lucacasonato) [reply pages/x/index.tsx:20]: There should be a test for it.

Comment (sandeshdamkondwar) [reply pages/x/index.tsx:20]: Nope, there is none. There is a prettier linting check, that's it.

Comment (lucacasonato) [reply pages/x/index.tsx:20]: https://github.com/denoland/deno_website2/blob/master/util/database.test.ts#L50-L53

File: pages/x/index.tsx (hunk 1 above)
Comment (rjoydip) [reply pages/x/index.tsx:21]: > Does it make sense to sort every time query changes? Sorting algorithms are expensive compared to filter. And sorting is one time operation when entries are changed. That's the idea behind separating.

Sorting don't need https://github.com/denoland/deno_website2/pull/1171#discussion_r443072926
It has test case [database.test.ts#L50-L53](https://github.com/denoland/deno_website2/blob/master/util/database.test.ts#L50-L53)

File: pages/x/index.tsx (hunk 2 above)
Comment (rjoydip) [reply pages/x/index.tsx]: @sandeshdamkondwar I don't have any solution right now. I will try to look solution later.

**Comment by ry** [comment]
I will wait for @lucacasonato's approval before landing

File: pages/x/index.tsx (hunk 3 above)
Comment (sandeshdamkondwar) [reply pages/x/index.tsx]: > https://github.com/denoland/deno_website2/blob/master/util/database.test.ts#L50-L53

Removing the sorting operation. Updated the PR.

File: pages/x/index.tsx (hunk 2 above)
Comment (sandeshdamkondwar) [reply pages/x/index.tsx]: Yes, will go that way. `a + ".*?" + b` is working fine. I'll revert back to that position. If you find any optimal solution, feel free to add in separate PR.

**Pull request merged**